metrics/sinks/statsd/etsystatsd_formatter.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statsd
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/heapster/metrics/core"
+	"sort"
+	"strings"
+)
+
+type EtsystatsdFormatter struct {
+	delimReplacer *strings.Replacer
+}
+
+func (formatter *EtsystatsdFormatter) Format(prefix string, name string, labels map[string]string, customizeLabel CustomizeLabel, metricValue core.MetricValue) (res string, err error) {
+	var metricName string
+	var suffix string
+	resourceId, ok := labels[core.LabelResourceID.Key]
+	if ok {
+		metricName = fmt.Sprintf("%s.%s",
+			formatter.delimReplacer.Replace(name),
+			formatter.delimReplacer.Replace(resourceId))
+	} else {
+		metricName = formatter.delimReplacer.Replace(name)
+	}
+	if prefix != "" {
+		prefix = formatter.delimReplacer.Replace(prefix) + "."
+	}
+	userLabelStr := labels[core.LabelLabels.Key]
+	if len(userLabelStr) > 0 {
+		suffix = fmt.Sprintf("%s.%s:%v|g",
+			formatter.formatUserLabels(userLabelStr, customizeLabel),
+			metricName,
+			metricValue.GetValue(),
+		)
+	} else {
+		suffix = fmt.Sprintf("%s:%v|g",
+			metricName,
+			metricValue.GetValue(),
+		)
+	}
+	if t, ok := labels[core.LabelMetricSetType.Key]; ok {
```

Comment:
move this so that it reads like
```golang
metricType, hasMetricType := labels[...]
if !hasMetricType {
    return "", fmt.Errorf(...)
}
switch metricType {
...
}
```